File: Algorigthm/2.2/Sort.js

var datas = [23,3,43,12,78,2,90,4,1,101,8,413,5,2];

// 冒泡排序*****************************************************************************************************************************
function bubbleSort(numbers) {
  var len = numbers.length;
  for (var i = 0; i < len; i++) {
    for (var j = 0; j < len - i - 1; j++) {
      if (numbers[j] > numbers[j+1]) {
        var temple = numbers[j];
        numbers[j] = numbers[j+1];
        numbers[j+1] = temple;
      }
    }
  }
  return numbers;
}

// 选择排序*****************************************************************************************************************************
function selectSort(numbers) {
  var len = numbers.length;

  for (var i = 0; i < len; i++) {
    for (var j = i + 1; j < len; j++) {
      if (numbers[i] > numbers[j]) {
        var temple = numbers[i];
        numbers[i] = numbers[j];
        numbers[j] = temple;
      }
    }
  }
  return numbers;
}

// 插入排序*****************************************************************************************************************************
function insertSort(numbers) {
  var len = numbers.length;

  for (var i = 1; i < len; i++) {
    var  value_i = numbers[i]
    var  j = i - 1;
    while (j > -1 && numbers[j] > value_i) {
      numbers[j + 1] = numbers[j--];
    }
    j++;
    numbers[j] = value_i;
    }
  return numbers;
}

// 希尔排序*****************************************************************************************************************************
function shellSort(numbers) {

  var len = numbers.length;
  var gaps = [5,3,1];

  for(var i = 0;i < gaps.length; i++) {
    var gap = gaps[i];

    for (var index = 0; index < len; index +=gap) {
      var value_index = numbers[index];
      var j = index - gap;

      while (j > -1 && numbers[j] > value_index) {
        numbers[j + gap] = numbers[j];
        j -= gap;
      }

      j +=gap;
      numbers[j] = value_index;
    }
  }
  return numbers;
}

// 快速排序*****************************************************************************************************************************

function division(numbers, left, right) {

  if (left > right)
    return;

  var len = numbers.length;

  var base = numbers[left];

  while (left < right) {

  while (left < right && numbers[right] > base) {
    right--;
  }
  numbers[left] = numbers[right];

  while (left < right && numbers[left] <= base) {
    left++;
  }
  numbers[right] = numbers[left];
  }

  numbers[left] = base;
  return left;
}

function quickSort(numbers,leftIndex,rightIndex) {
  if (leftIndex < rightIndex) {
    var  mid = division(numbers, leftIndex , rightIndex);
    quickSort(numbers, leftIndex, mid - 1);
    quickSort(numbers, mid + 1, rightIndex);
  }
}

// 归并排序 ***************************************************************************************************************************
function mergeSort(numbers) {

  function min (a, b) {
    return a < b ? a : b;
  }

  var len = numbers.length;
  var teamLength = 1;
  var b = [];
  var k = 0;

  for (teamLength = 1; teamLength < len; teamLength += teamLength) {
    k = 0;

    for (var index = 0; index < len; index += teamLength + teamLength) {

      var low1 = index;
      var end1 = min(index + teamLength, len);

      var low2 = end1 ;
      var end2 = min(low2 + teamLength, len);

      while (low1 < end1 && low2 < end2) {
        b[k++] = numbers[low1] < numbers[low2] ? numbers[low1++] : numbers[low2++];
      }

      while (low1 < end1) {
        b[k++] = numbers[low1++];
      }

      while (low2 < end2) {
        b[k++] = numbers[low2++];
      }
    }
    for (var i = 0; i < len; i++) {
      numbers[i] = b[i];
    }
  }
  return numbers;

}

// 桶排序 *****************************************************************************************************************************
function bucketSort(numbers) {


}

// // 堆排序 *****************************************************************************************************************************
// function  heapSort() {
//
// }

// 测试排序
console.log("元数据   :" + datas+ "\n");

var result;

// var result = bubbleSort(datas);
// var result = selectSort(datas);
// var result = insertSort(datas);
// var result = shellSort(datas);
// result = quickSort(datas,0,datas.length - 1);
result = mergeSort(datas);

console.log("排列结果 :" + datas);

File: Algorigthm/3.7/Algorithm_Sort/Algorithm_Sort/Sort.swift
//
//  Sort.swift
//  Algorithm_Sort
//
//  Created by ssb on 3/7/18.
//  Copyright © 2018 ssb. All rights reserved.
//

import Cocoa

// 排序算法 从小到大的顺序

/*
冒泡排序
选择排序
插入排序
快速排序
希尔排序
归并排序
堆排序
*/

class Sort: NSObject {
    
    // 模拟数据
    var testNumbers: [Int] = [2,45,33,12,31,89,1,5,80,52]

    // 外部执行
    func execute() -> Void {
    var result: [Int]?

    print("original data: \n \(testNumbers)")

//        result = self.bubbleSort(numbers: self.testNumbers)
//        result = selectSort(numbers: self.testNumbers)
//        result = insertSort(numbers: self.testNumbers)
//    result = quickSort(numbers: self.testNumbers)
//        result = shellSort(numbers: self.testNumbers)
//        result  = mergeSort(numbers: self.testNumbers)
        result = heapSort(numbers: self.testNumbers)
    print("results is :\n \(result!)")
}

// MARK: - 冒泡排序
fileprivate  func bubbleSort(numbers: [Int]) -> [Int] {
    var numbers2 = numbers
    
    for index in 0..<(numbers2.count - 1) {
        var index_j: Int = 0;
        
        while index_j < numbers2.count - index - 1 {
            
            if numbers2[index_j] > numbers2[index_j + 1] {
                let temp = numbers2[index_j]
                numbers2[index_j] = numbers2[index_j + 1]
                numbers2[index_j + 1] = temp
            }
            index_j += 1;
        }
    }
    return numbers2
}

// MARK: - 选择排序
private func selectSort(numbers: [Int]) -> [Int] {
    var numbers2 = numbers
    
    for index in 0..<numbers2.count {
        var index_next = index + 1
        
        while index_next < numbers2.count {
            if numbers2[index] > numbers2[index_next] {
                let temp = numbers2[index]
                numbers2[index] = numbers2[index_next]
                numbers2[index_next] = temp
            }
            index_next += 1;
        }
    }
    return numbers2
}

// MARK: - 插入排序
private func insertSort(numbers: [Int]) -> [Int] {
    var numbers2 = numbers
    
    for index in 1..<numbers2.count{
        let value = numbers2[index]
        var beforeIndex = index - 1
        
        while beforeIndex > -1 && numbers2[beforeIndex] > value{
            numbers2[beforeIndex + 1] = numbers2[beforeIndex]
            beforeIndex -= 1;
        }
        
        beforeIndex += 1;
        numbers2[beforeIndex] = value
    }
    return numbers2;
}

    // MARK: - 快速排序
private func quickSort(numbers: [Int]) -> [Int] {
    var numbers2 = numbers;
    
    // 内部迭代函数
    func subQuickSort(startIndex: Int, endIndex: Int) {
        if startIndex < endIndex {
            let midIndex = division(startIndex: startIndex, endIndex: endIndex)
            subQuickSort(startIndex: startIndex, endIndex: midIndex - 1)
            subQuickSort(startIndex: midIndex + 1, endIndex: endIndex)
        }
    }

    // 切分
    func division(startIndex: Int, endIndex: Int) -> Int {
        if startIndex > endIndex {
            return numbers2.count + 100
        }
        
        var left = startIndex
        var right = endIndex
        let comparedValue = numbers2[startIndex]
        
        while left < right {
            while left < right && numbers2[right] > comparedValue {
                    right -= 1
                }
                numbers2[left] = numbers2[right]
            
            while left < right && numbers2[left] < comparedValue {
                    left += 1
                }
                numbers2[right] = numbers2[left]
            }
        
            numbers2[left] = comparedValue
            return left
    }
    
    subQuickSort(startIndex: 0, endIndex: numbers.count - 1)
    return numbers2;
    
    }
    // MARK: -  希尔排序
    func shellSort(numbers: [Int]) -> [Int] {
        var numbers2 = numbers
        var gap = (Int)(numbers2.count/2)
        
        while gap > 0 {
            print("gap is: \(gap)")
            var index = gap
            while index < numbers2.count {
                print(numbers2)
                let value = numbers2[index]
                var j = index
                
                while j >= gap && numbers2[j - gap] > value {
                    numbers2[j] = numbers2[j - gap]
                    j -= gap
                }
                numbers2[j] = value
                index += 1;
            }
            gap = (Int)(gap/2)
        }
        return numbers2
    }
    
    // MARK: - 归并排序
    func mergeSort(numbers: [Int]) -> [Int] {
        
        // 求较小值
        func min(a: Int, b: Int) -> Int {
            return a < b ? a : b
        }
        
        var unorderedNumbers = numbers
        var b = Array(repeatElement(0, count: unorderedNumbers.count));
        var step = 1
        var k = 0
        while step < unorderedNumbers.count {
            k = 0
            var index = 0
            
            while index < unorderedNumbers.count {
                var start1 = index
                let end1 = min(a: start1 + step, b: unorderedNumbers.count)
                
                var start2 = end1
                let end2 = min(a: start2 + step, b: unorderedNumbers.count)
                
                while start1 < end1 && start2 < end2 {
                    if unorderedNumbers[start1] < unorderedNumbers[start2] {
                        b[k] = unorderedNumbers[start1]
                        start1 += 1
                    } else {
                        b[k] = unorderedNumbers[start2]
                        start2 += 1
                    }
                    k += 1
                }
                
                while start1 < end1 {
                    b[k] = unorderedNumbers[start1]
                    k += 1
                    start1 += 1
                }
                
                while start2 < end2 {
                    b[k] = unorderedNumbers[start2]
                    k += 1
                    start2 += 1
                }
                index = index + step + step
            }
            for i in b.indices {
                unorderedNumbers[i] = b[i]
            }
            step *= 2
        }
    
        return unorderedNumbers;
    }
    
    // MARK: - 堆排序
    func heapSort(numbers: [Int]) -> [Int] {
        var size = numbers.count
        var array = numbers;
        
        // 调整堆
        func adjustHeap(s: Int, length: Int) {
            var index = s
            var j: Int, k: Int, max: Int
            while 2 * index + 1 < length {
                j = 2 * index + 1
                max = j
                if j + 1 < length {
                    if array[j] < array[j+1] {
                        max = j + 1
                    }
                }
                
                if array[index] < array[max] {
                    k = array[index]
                    array[index] = array[max]
                    array[max] = k
                    index = max
                } else {
                    break
                }
            }
        }
        
        var i = size/2
        while i >= 0 {
            adjustHeap(s: i, length: size)
            i -= 1
        }
        
        i = size - 1
        while i > 0 {
            let temple = array[0]
            array[0] = array[i]
            array[i] = temple
            adjustHeap(s: 0, length: i)
            i -= 1
            
        }
        return array
    }

}
